src/parser.rs

```diff
@@ -132,103 +132,198 @@ pub fn parse_lcov<T: Read>(
+                        iter.take_while(|&&c| c != b'\n').fold(0, |_, _| 0);
+                        continue;
                     }
-                    cur_functions.insert(
-                        f_name.to_owned(),
-                        Function {
-                            start,
-                            executed: false,
-                        },
-                    );
-                }
-                "FNDA" => {
-                    let mut f_splits = value.splitn(2, ',');
-                    let executed = try_next!(f_splits, l) != "0";
-                    let f_name = try_next!(f_splits, l);
-                    if let Some(f) = cur_functions.get_mut(f_name) {
-                        f.executed |= executed;
-                    } else {
-                        return Err(ParserError::Parse(format!(
-                            "FN record missing for function {}",
-                            f_name
-                        )));
-                    }
-                }
-                "BRDA" => {
-                    if branch_enabled {
-                        let mut values = value.splitn(4, ',');
-                        let line_no = try_parse_next!(values, l);
-                        values.next();
-                        let branch_number = try_parse_next!(values, l);
-                        let taken = try_next!(values, l) != "-";
-                        add_branch(&mut cur_branches, line_no, branch_number, taken);
+
+                    let key = iter
+                        .take_while(|&&c| c != b':')
+                        .fold(*c as u32, |r, &x| r * (1 << 8) + u32::from(x));
+                    match key {
+                        SF => {
+                            // SF:string
+                            cur_file = Some(iter.take_while(|&&c| c != b'\n' && c != b'\r').fold(
+                                String::new(),
+                                |mut b, &c| {
+                                    b.push(c as char);
+                                    b
+                                },
+                            ));
+                        }
+                        DA => {
+                            // DA:uint,int
+                            let line_no = iter
+                                .take_while(|&&c| c != b',')
+                                .fold(0, |r, &x| r * 10 + u32::from(x - b'0'));
+                            if iter.peek().is_none() {
+                                return Err(ParserError::InvalidRecord(format!(
+                                    "DA at line {}",
+                                    line
+                                )));
+                            }
+                            let execution_count = if let Some(c) = iter.next() {
+                                if *c == b'-' {
+                                    iter.take_while(|&&c| c != b'\n').fold(0, |_, _| 0)
```

Comment:
Instead of `.fold(0, |_, _| 0)` to consume iterators, it might be clearer to use `count()` or `last()`